tutorials/01-gaussian-mixture-model/01_gaussian-mixture-model.jmd

```diff
@@ -186,8 +242,168 @@ scatter(
+
+In fact, `Distributions.jl` provides [many convenient constructors](https://juliastats.org/Distributions.jl/stable/mixture/) for mixture models, allowing further simplification in common special cases (mixtures distributions are of the same type, weight vector is known to allocate probability equally, and so on).
+
+The `logpdf` implementation for a `MixtureModel` distribution is exactly the marginalization defined above, and so our model becomes simply:
+
+```julia
+@model function gmm_marginalized(x)
+    K = 2
+    D, _ = size(x)
+    μ ~ ordered(MvNormal(Zeros(K), I))
+    w ~ Dirichlet(K, 1.0)
+
+    x ~ MixtureModel([MvNormal(Fill(μₖ, D), I) for μₖ in μ], w)
+end
+
+model = gmm_marginalized(x);
+```
+
+As we've summed out the discrete components, we can perform inference using `NUTS()` alone. 
+
+```julia
+sampler = NUTS()
+chains = sample(model, sampler, MCMCThreads(), nsamples, nchains; discard_initial = burn);
+```
+
+```julia; echo=false
+let
+    # Verify for marginalized model that the output of the chain is as expected
+    for i in MCMCChains.chains(chains)
+        # μ[1] and μ[2] can no longer switch places. Check that they've found the mean
+        chain = Array(chains[:, ["μ[1]", "μ[2]"], i])
+        μ_mean = vec(mean(chain; dims=1))
+        @assert isapprox(sort(μ_mean), μ; rtol=0.4) "Difference between estimated mean of μ ($(sort(μ_mean))) and data-generating μ ($μ) unexpectedly large!"
+    end
+end
+```
+
+`NUTS()` significantly outperforms our compositional Gibbs sampler, in large part because our model is now Rao-Blackwellized thanks to 
+the marginalization of our assignment parameter.
+
+
+```julia
+plot(chains[["μ[1]", "μ[2]"]], legend=true)
+```
+
+## Inferred Assignments - Marginalized Model
+
+As we've summed over possible assignments, the associated parameter is no longer available in our chain. 
+This is not a problem, however, as given any fixed sample $(\mu, w)$, the assignment probability — $p(z_i \mid y_i)$ — can be recovered using Bayes rule:
+
+$$
+p(z_i \mid y_i) = \frac{p(y_i \mid z_i) p(z_i)}{\sum_{k = 1}^K \left(p(y_i \mid z_i) p(z_i) \right)}
+$$
+
+This quantity can be computed for every $p(z = z_i \mid y_i)$, resulting in a probability vector, which is then used to sample 
+posterior predictive assignments from a categorial distribution.
+
+For details on the mathematics here, see [the Stan documentation on latent discrete parameters](https://mc-stan.org/docs/stan-users-guide/latent-discrete.html).
+
+```julia
+@model function gmm_recover(x)
+    K = 2
+    D, N =  size(x)
+    μ ~ ordered(MvNormal(Zeros(K), I))
+    w ~ Dirichlet(K, 1.0)
+
+    dists = [MvNormal(Fill(μₖ, D), I) for μₖ in μ]
+
+    x ~ MixtureModel(dists, w)
+
+    # Return sample_class(yi) for fixed μ, w.
+    function sample_class(xi)
+        lvec = [(logpdf(d, xi) + log(w[i])) for (i, d) in enumerate(dists)]
+        rand(Categorical(exp.(lvec .- logsumexp(lvec))))
+    end
```

Comment:
This should be defined outside of the model and probably use `softmax` or `softmax!` directly.